src/cgutils.cpp

```diff
@@ -1165,10 +1168,13 @@ static Value *emit_bounds_check(const jl_cgval_t &ainfo, jl_value_t *ty, Value *
 // It is currently unused, but might be used in the future for a more precise answer.
 static unsigned julia_alignment(Value* /*ptr*/, jl_value_t *jltype, unsigned alignment)
 {
-    if (!alignment && ((jl_datatype_t*)jltype)->layout->alignment > MAX_ALIGN) {
-        // Type's natural alignment exceeds strictest alignment promised in heap, so return the heap alignment.
-        return MAX_ALIGN;
+    if (!alignment) {
+        alignment = jl_datatype_align(jltype);
     }
+    if (alignment > JL_HEAP_ALIGNMENT)
```

Comment:
This should be an assertion in the condition above. When used in `pointerref`/`pointerset` the user should be allowed to specify larger alignment.